File: game/src/main/java/me/firecloud/game/rule/ActionNode.java
/**
 * 
 */
package me.firecloud.game.rule;

/**
 * @date 2012-10-12
 * 
 */
public abstract class ActionNode
    extends SingleNode
    implements Node {

  /*
   * (non-Javadoc)
   * 
   * @see me.firecloud.game.rule.Node#handle(me.firecloud.game.rule.Event)
   */
  public
      Node handle(Event event,GameContext gameContext) throws RuleException {
    if (event.isHandled()) {
      return this;
    } else {
      if (perform(event,gameContext)) {
        //
        return nextNode.handle(event,gameContext);
      } else {
        return null;
      }
    }
  }

  /**
   * 
   * @param event
   * @return
   * @throws RuleException
   */
  protected abstract
      boolean perform(Event event,GameContext gameContext) throws RuleException;

}

File: card/src/main/scala/me/firecloud/card/rule/ActionNode.scala
/**
 *
 */
package me.firecloud.card.rule

/**
 * @author kkppccdd
 * @email kkppccdd@gmail.com
 * @date Oct 19, 2012
 *
 */
abstract class ActionNode(nextNode:Node) extends SingleNode(nextNode:Node) {

  override def handle(gameContext: GameContext, event: Event): Node = {
    if (event.isHandled) {
      return null;
    } else {
      if(perform(gameContext, event)){
        return nextNode.handle(gameContext,event)
      }else{
        return null
      }
    }
  }

   protected def perform(gameContext: GameContext, event: Event):Boolean

}